pkg/controller/petset/pet_set.go

```diff
@@ -0,0 +1,342 @@
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the pod
+	// changed labels the new PetSet will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*api.Pod)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not a pod %+v", obj)
+			return
+		}
+	}
+	glog.Infof("Pod %s/%s deleted through %v.", pod.Namespace, pod.Name, utilruntime.GetCaller())
+	if ps := psc.getPetSetForPod(pod); ps != nil {
+		psc.enqueuePetSet(ps)
+	}
+}
+
+func (psc *PetSetController) getPodsForPetSet(ps *apps.PetSet) ([]*api.Pod, error) {
+	// TODO: Do we want the petset to fight with RCs? check parent petset annoation, or name prefix?
+	sel, err := unversioned.LabelSelectorAsSelector(ps.Spec.Selector)
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	petList, err := psc.podStore.Pods(ps.Namespace).List(sel)
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	pods := []*api.Pod{}
+	for _, p := range petList.Items {
+		pods = append(pods, &p)
+	}
+	return pods, nil
+}
+
+// getPetSetForPod returns the pet set managing the given pod.
+func (psc *PetSetController) getPetSetForPod(pod *api.Pod) *apps.PetSet {
+	ps, err := psc.psStore.GetPodPetSets(pod)
+	if err != nil {
+		glog.V(4).Infof("No PetSets found for pod %v, PetSet controller will avoid syncing", pod.Name)
+		return nil
+	}
+	// Resolve a overlapping petset tie by creation timestamp.
+	// Let's hope users don't create overlapping petsets.
+	if len(ps) > 1 {
+		glog.Errorf("user error! more than one PetSet is selecting pods with labels: %+v", pod.Labels)
+		sort.Sort(overlappingPetSets(ps))
+	}
+	return &ps[0]
+}
+
+func (psc *PetSetController) enqueuePetSet(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Cound't get key for object %+v: %v", obj, err)
+		return
+	}
+	psc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (psc *PetSetController) worker() {
+	for {
+		func() {
+			key, quit := psc.queue.Get()
+			if quit {
+				return
+			}
+			defer psc.queue.Done(key)
+			if errs := psc.syncHandler(key.(string)); len(errs) != 0 {
+				glog.Errorf("Error syncing PetSet, requeuing: %v", errs)
+				psc.queue.Add(key)
+			}
+		}()
+	}
+}
+
+// Sync syncs the given petset.
+func (psc *PetSetController) Sync(key string) []error {
+	startTime := time.Now()
+	defer func() {
+		glog.Infof("Finished syncing pet set %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	if !psc.podStoreSynced() {
+		// Sleep so we give the pod reflector goroutine a chance to run.
+		time.Sleep(PodStoreSyncedPollPeriod)
```

Comment:
obviously, it's the pattern we follow in other controllers and fits the purpose, insted of overengineering a yield in there.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L506